File: components/audit/src/main/java/org/cloudfoundry/credhub/audit/RequestDetails.java
package org.cloudfoundry.credhub.audit;

import java.io.IOException;

import com.fasterxml.jackson.databind.ObjectMapper;

import static com.fasterxml.jackson.databind.SerializationFeature.FAIL_ON_EMPTY_BEANS;

public interface RequestDetails {
  default String toJSON() {
    final String result;
    try {
      final ObjectMapper mapper = new ObjectMapper();
      mapper.configure(FAIL_ON_EMPTY_BEANS, false);
      result = mapper.writeValueAsString(this);
    } catch (final IOException e) {
      throw new RuntimeException(e);
    }

    return result;
  }

  OperationDeviceAction operation();
}

File: components/audit/src/main/kotlin/org/cloudfoundry/credhub/audit/RequestDetails.kt
package org.cloudfoundry.credhub.audit

import tools.jackson.databind.SerializationFeature.FAIL_ON_EMPTY_BEANS
import tools.jackson.databind.json.JsonMapper

interface RequestDetails {
    fun toJSON(): String {
        val mapper =
            JsonMapper
                .builder()
                .disable(FAIL_ON_EMPTY_BEANS)
                .build()
        return mapper.writeValueAsString(this)
    }

    fun operation(): OperationDeviceAction
}
